### ISSUE TITLE
Server-side mutations/queries

### ISSUE THREAD

**Issue by TimoBechtel** [TimoBechtel/socketdb#47]
# Server Mutations / Queries

The server side api is currently quite hard to use and very low-level.
The following are a few very rough ideas on how to improve the server-side api, especially for data validation.

## Server example
```ts
const playerRoute = route({
	create: procedure
		.arguments(z.object({
			name: z.string()
		}))
		.mutate(({args, db}) => {
			db.get('players').get(nanoid()).set(args);
		}),
	move: procedure
		.arguments(z.string())
		// multiple payloads will be batched
		.payload(z.object({
			x: z.number(),
			y: z.number()
		}))
		// call with same arguments will only be sent again when local cache changes, at the given path
		.affects(({args}) => [`/players/${args}/x`, `/players/${args}/y`])
		.mutate(({args, payload, db}) => {
			db.get('players').get(args).set(payload);
		}),
	getWithinRange: procedure
		.arguments(z.object({
			x: z.number(),
			y: z.number(),
			width: z.number(),
			height: z.number(),
		}))
		// calls to "db" are automatically tracked to create a dependencies array
		// whenever any of the dependencies change, the query is re-run
		.query(({args, db}) => {
			return db.get('players').values().filter(user => isWithinBoundingBox(args, user));
		}),
	getTeamsForPlayer: procedure
		.arguments(z.string())
		.query(({args, db}) => {
			const player = db.get('players').get(args);
			return db.get('teams').value().filter(team => team.id === player.teamId);
		}),
	// Theoretically, we could also allow to define events here
	poke: procedure
		.arguments(z.object({
			userId: z.string(),
			message: z.string()
		}))
		.mutate(({args, api}) => {
			if (!allowedToPoke(args.userId)) throw new Error('Not allowed to poke');
			// if the user is subscribed to this channel, the user will receive the message
			api.getClient(({userId}) => userId === args.userId).send('poke', args.message);
		}),
	// ....
});

const router = createRouter({
	player: userRoute
});
```

## This can then be used like:

```ts
const client = createClient(router);

client.player.create({name: 'Michael'});
client.player.create({name: 'Michael'});
// sends:
// [
// 	["player/create", {name: 'Michael'}],
// 	["player/create", {name: 'Michael'}]
// ]

client.player.move('1', {x: 0, y: 1});
client.player.move('1', {x: 2, y: 1});
client.player.move('1', {x: 2, y: 1});
client.player.move('2', {x: 5, y: 10});
// sends:
// [
// 	["/player/move", '1', {x: 2, y: 1}],
// 	["/player/move", '2', {x: 5, y: 10}],
// ]

client.player.getTeamsForPlayer('1').on(teams => console.log(teams));
// => []
client.player.addToTeam({id: '1', teamId: '2'});
// => [{id: 2}]
```

## A few notes on how this works:
There are a) **mutations** and b) **queries**.

### Mutations
Mutate values on the server.
- `arguments()`: define the input to the function, the given schema validates the input from the client
- `payload()`:  the same as `arguments()`, but this will be batched on consecutive calls
- `affects()`: allows you to define paths to check for changes before calling the function with the same arguments
   - this will skip multiple calls with the same arguments / payload when the data for the given paths has not changed
   - for example: the player sends a position update 60x per second. But we do not need to send an update if the position has not changed. We cannot run the mutation on the client, but we can skip consecutive calls with the same arguments/payload. As long as we know that the data (defined by `affects()`) has not been modified by another client.
- `mutate()`: runs the mutation

### Queries
In addition to the above:
- `query()`: queries the database and returns the result.
   - Calls to the database (e.g. `db.get('players')` will be tracked. Anytime those dependencies change, the query will re-run and notify the client.

## I think this solves multiple issues:
- allows defining strict routes on the server
- includes validation out of the box
- provides a flexible way to write server-side logic (server-side functions, filter/queries, events, etc.)
- adds a powerful api to create reactive queries
- fixes the problem of creating ids on the client side (e.g. adding things to lists)
- allows for easy authorization logic
- completely gets rid of clumsy update interceptors (e.g. `intercept('server:update')`)
- still allows to batch updates: multiple calls with the same arguments but different payloads will be batched into a single call with combined payloads
- only sends updates when needed (by defining `affects`)
- support for other store implementations
   - we could for example wrap things like prisma, allowing us to directly query a database

## Shortcomings
- The api is a bit confusing:
   - I think the difference between `arguments()` and `payload()` might not be very clear.
   - Some functions, like `affects` and `payload` do not have an effect on queries. So maybe the api should be a bit different and not allow them to be chained.
   - The user has to define affected paths to make the mutation more efficient. This might be error-prone and a bit verbose.

**Comment by TimoBechtel** [comment]
# A different concept, heavily inspired by tRPC:

## Server example
```ts
const playerRoute = route({
	create: procedure
		.arguments(z.object({
			name: z.string()
		}))
		.mutate(({args, db}) => {
			db.get('players').get(nanoid()).set(args);
		}),
	move: procedure
		.arguments(z.string())
		// multiple payloads will be batched
		.payload(z.object({
			x: z.number(),
			y: z.number()
		}))
		// call with same arguments will only be sent again when local cache changes, at the given path
		.affects(({args}) => [`/players/${args}/x`, `/players/${args}/y`])
		.mutate(({args, payload, db}) => {
			db.get('players').get(args).set(payload);
		}),
	getWithinRange: procedure
		.arguments(z.object({
			x: z.number(),
			y: z.number(),
			width: z.number(),
			height: z.number(),
		}))
		// calls to "db" are automatically tracked to create a dependencies array
		// whenever any of the dependencies change, the query is re-run
		.query(({args, db}) => {
			return db.get('players').filter(user => isWithinBoundingBox(args, user));
		}),
	getTeamsForPlayer: procedure
		.arguments(z.string())
		.query(({args, db}) => {
			const player = db.get('players').get(args);
			return db.get('teams').filter(team => team.id === player.teamId);
		}),
	// ....
});

const router = createRouter({
	player: userRoute
});
```

## This can then be used like:

```ts
const client = createClient(router);

client.player.create({name: 'Michael'});
client.player.create({name: 'Michael'});
// sends:
// [
// 	["player/create", {name: 'Michael'}],
// 	["player/create", {name: 'Michael'}]
// ]

client.player.move('1', {x: 0, y: 1});
client.player.move('1', {x: 2, y: 1});
client.player.move('1', {x: 2, y: 1});
client.player.move('2', {x: 5, y: 10});
// sends:
// [
// 	["/player/move", '1', {x: 2, y: 1}],
// 	["/player/move", '2', {x: 5, y: 10}],
// ]

client.player.getTeamsForPlayer('1').on(teams => console.log(teams));
// => []
client.player.addToTeam({id: '1', teamId: '2'});
// => [{id: 2}]
```

## A few notes on how this works:
There are a) **mutations** and b) **queries**.

### Mutations
Mutate values on the server.
- `arguments()`: define the input to the function, the given schema validates the input from the client
- `payload()`:  the same as `arguments()`, but this will be batched on consecutive calls
- `affects()`: allows you to define paths to check for changes before calling the function with the same arguments
   - this will skip multiple calls with the same arguments / payload when the data for the given paths has not changed
   - for example: the player sends a position update 60x per second. But we do not need to send an update if the position has not changed. We cannot run the mutation on the client, but we can skip consecutive calls with the same arguments/payload. As long as we know that the data (defined by `affects()`) has not been modified by another client.
- `mutate()`: runs the mutation

### Queries
In addition to the above:
- `query()`: queries the database and returns the result.
   - Calls to the database (e.g. `db.get('players')` will be tracked. Anytime those dependencies change, the query will re-run and notify the client.

## I think this solves multiple issues:
- allows defining strict routes on the server
- includes validation out of the box
- provides a flexible way to write server-side logic (server-side functions, filter/queries, events, etc.)
- adds a powerful api to create reactive queries
- fixes the problem of creating ids on the client side (e.g. adding things to lists)
- allows for easy authorization logic
- completely gets rid of clumsy update interceptors (e.g. `intercept('server:update')`)
- still allows to batch updates: multiple calls with the same arguments but different payloads will be batched into a single call with combined payloads
- only sends updates when needed (by defining `affects`)

## Shortcomings
- The api is a bit confusing:
   - I think the difference between `arguments()` and `payload()` might not be very clear.
   - Some functions, like `affects` and `payload` do not have an effect on queries. So maybe the api should be a bit different and not allow them to be chained.
   - The user has to define affected paths to make the mutation more efficient. This might be error-prone and a bit verbose.

**Comment by TimoBechtel** [comment]
Addressed by #79
